File: KingPastaKit/app/schedule.rb
class Schedule
  include MotionModel::Model
  include MotionModel::ArrayModelAdapter
  include MotionModel::Validatable

  columns :name => { :type => :string, :default => "" },
          :showsNotifications => { :type => :boolean, :default => false }

  # validates :name, :presence => true
  # validates :showsNotifications, :presence => true

  has_many :days
  has_many :periods

  def ordered_periods
    self.periods.order(:startTime)
  end

  def days_string
    case days.count
    when 0
      " "       # space stops detailTextLabel collapsing to a size of 0 in the ScheduleController
    when 1
      days.first.name
    when 2
      days_array = days.order(:dayOfWeek).array
      days_array = days_array.reverse if days_array[0].dayOfWeek == 0 && days_array[1].dayOfWeek == 6   # puts days 0 & 6 in running order
      days_array.map { |day| day.name }.join(" & ")
    else
      days_array = days.order(:dayOfWeek).array
      days_array.take(days_array.count - 1).map { |day| Day.shortSymbols[day.dayOfWeek] }.join(", ") + " & #{Day.shortSymbols[days_array.last.dayOfWeek]}"
    end
  end

  def started?
    return false if ordered_periods.count < 1
    Time.now.strip_date > ordered_periods.array.first.startTime
  end

  def start_time
    case self.periods.count
    when 0
      nil
    when 1
      self.periods.first.startTime
    else
      self.ordered_periods.array.reject { |p| p.startTime == nil }.first.startTime  # rejects unsaved objects which don't have a start time
    end
  end

  def end_time
    case self.periods.count
    when 0
      nil
    when 1
      self.periods.first.endTime
    else
      self.ordered_periods.array.reject { |p| p.startTime == nil }.last.endTime  # array helps reliable objects while editing and allows for .last
    end
  end

  def length
    end_time - start_time
  end

  def shows_notifications?
    self.showsNotifications unless self.showsNotifications.nil?
  end

  def shows_notifications=(bool)
    self.showsNotifications = bool
  end

  def self.today
    if day_today = Day.today then day_today.schedule end
  end

  def self.tomorrow
    if day_tomorrow = Day.tomorrow then day_tomorrow.schedule end
  end

  def self.on_wday(wday)
    if day_wday = Day.wday(wday) then day_wday.schedule end
  end

end

File: KingPastaKit/Schedule.swift
import Foundation
import RealmSwift

class Schedule: Object {
    dynamic var name = ""
    dynamic var showsNotifications = false
    
    var days: Results<Day> {
        return FkP.realm.objects(Day).filter("schedule == %@", self)
    }
    var periods: Results<Period> {
        return FkP.realm.objects(Period).filter("schedule == %@", self)
    }
    
    func orderedPeriods() -> Results<Period> {
        return periods.sorted("startTime")
    }
    
    func daysString() -> String {
        switch days.count {
        case 0:
            return " "       // space stops detailTextLabel collapsing to a size of 0 in the ScheduleController
        case 1:
            return days.first!.name
        case 2:
            var daysArray = Array(days.sorted("dayOfWeek"))
            // puts days 0 & 6 in running order
            if daysArray[0].dayOfWeek == 0 && daysArray[1].dayOfWeek == 6 {
                daysArray = daysArray.reverse()
            }
            return daysArray.map { $0.name }.joinWithSeparator(" & ")
        default:
            let daysArray = Array(days.sorted("dayOfWeek"))
            return daysArray.prefix(daysArray.count - 1).map { Day.shortSymbols[$0.dayOfWeek] }.joinWithSeparator(", ") + " & \(Day.shortSymbols[daysArray.last!.dayOfWeek])"
        }
    }
    var hasStarted: Bool {
        if orderedPeriods().count < 1 {
            return false
        } else if let firstPeriod = orderedPeriods().first {
            return NSDate().stripDate().compare(firstPeriod.startTime) == .OrderedDescending // >
        }
        // TODO: Better swift flow?
        return false
    }
    
    func startTime() -> NSDate? {
        switch periods.count {
        case 0:
            return nil
        case 1:
            return periods.first!.startTime
        default:
            // rejects unsaved objects which don't have a start time
            return orderedPeriods().filter { $0.startTime != nil }.first?.startTime
        }
    }
    
    func endTime() -> NSDate? {
        switch periods.count {
        case 0:
            return nil
        case 1:
            return periods.first?.endTime
        default:
            // RubyMotion: array helps reliable objects while editing and allows for .last
            return orderedPeriods().filter { $0.startTime != nil }.last?.endTime
        }
    }
    
    func length() -> NSDate? {
        guard let endTime = endTime(), startTime = startTime() else { return nil }
        return NSDate(timeIntervalSince1970: endTime.timeIntervalSinceDate(startTime))
    }
    
    class func today() -> Schedule? {
        guard let dayToday = Day.today() else { return nil }
        return dayToday.schedule
    }
    
    class func tomorrow() -> Schedule? {
        guard let dayTomorrow = Day.tomorrow() else { return nil }
        return dayTomorrow.schedule
    }
    
    class func onWday(wday: Int) -> Schedule? {
        guard let dayWday = Day.wday(wday) else { return nil }
        return dayWday.schedule
    }
}
